File: RocketChat.Client/RocketChatApi.cs
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.IO;
using System.Text;
using System.Text.Json;

namespace RocketChat.Client;

public sealed class RocketChatApi : IDisposable
{
    private readonly HttpClient _http = new() { Timeout = TimeSpan.FromSeconds(30) };
    private readonly JsonSerializerOptions _json = new() { PropertyNameCaseInsensitive = true };
    public Uri? ServerUri { get; private set; }
    public Session? Session { get; private set; }

    public async Task<(string Version, string Name)> GetServerInfoAsync(string server, CancellationToken ct = default)
    {
        ConfigureServer(server);
        using var doc = await SendAsync(HttpMethod.Get, "api/info", null, false, ct);
        var info = doc.RootElement.TryGetProperty("info", out var i) ? i : doc.RootElement;
        return (GetString(info, "version"), GetString(info, "name", ServerUri!.Host));
    }

    public async Task<Session> LoginAsync(string server, string username, string password, CancellationToken ct = default)
    {
        ConfigureServer(server);
        using var doc = await SendAsync(HttpMethod.Post, "api/v1/login", new { user = username, password }, false, ct);
        var data = doc.RootElement.GetProperty("data");
        var me = data.GetProperty("me");
        Session = new(GetString(data, "userId"), GetString(data, "authToken"), GetString(me, "username", username), GetString(me, "name", username));
        return Session;
    }

    public async Task LogoutAsync(CancellationToken ct = default)
    {
        if (Session is null) return;
        try { using var _ = await SendAsync(HttpMethod.Post, "api/v1/logout", new { }, true, ct); }
        finally { Session = null; }
    }

    public async Task<IReadOnlyList<Room>> GetRoomsAsync(CancellationToken ct = default)
    {
        using var doc = await SendAsync(HttpMethod.Get, "api/v1/subscriptions.get", null, true, ct);
        var root = doc.RootElement;
        var array = root.TryGetProperty("update", out var update) ? update : root.TryGetProperty("subscriptions", out var subscriptions) ? subscriptions : default;
        if (array.ValueKind != JsonValueKind.Array) return [];
        return array.EnumerateArray().Where(x => GetBool(x, "open", true)).Select(x => new Room
        {
            Id = GetString(x, "rid"), Name = GetString(x, "name"), DisplayName = GetString(x, "fname", GetString(x, "name")),
            Type = GetString(x, "t", "c"), Unread = GetInt(x, "unread"), IsFavorite = GetBool(x, "f"), IsOpen = GetBool(x, "open", true),
            LastMessageAt = GetDate(x, "ls")
        }).Where(x => x.Id.Length > 0).OrderByDescending(x => x.IsFavorite).ThenByDescending(x => x.LastMessageAt).ToList();
    }

    public async Task<IReadOnlyList<ChatMessage>> GetHistoryAsync(Room room, int count = 50, DateTimeOffset? latest = null, CancellationToken ct = default)
    {
        var endpoint = room.Type switch { "d" => "im.history", "p" => "groups.history", _ => "channels.history" };
        var query = $"api/v1/{endpoint}?roomId={E(room.Id)}&count={Math.Clamp(count, 1, 100)}" + (latest is null ? "" : $"&latest={E(latest.Value.UtcDateTime.ToString("O"))}");
        using var doc = await SendAsync(HttpMethod.Get, query, null, true, ct);
        if (!doc.RootElement.TryGetProperty("messages", out var messages)) return [];
        return messages.EnumerateArray().Select(ParseMessage).OrderBy(x => x.Timestamp).ToList();
    }

    public async Task<ChatMessage> SendMessageAsync(string roomId, string text, string? threadId = null, CancellationToken ct = default)
    {
        var message = threadId is null ? new Dictionary<string, object?> { ["rid"] = roomId, ["msg"] = text }
            : new Dictionary<string, object?> { ["rid"] = roomId, ["msg"] = text, ["tmid"] = threadId };
        using var doc = await SendAsync(HttpMethod.Post, "api/v1/chat.sendMessage", new { message }, true, ct);
        return ParseMessage(doc.RootElement.GetProperty("message"));
    }

    public async Task UpdateMessageAsync(string roomId, string messageId, string text, CancellationToken ct = default)
    { using var _ = await SendAsync(HttpMethod.Post, "api/v1/chat.update", new { roomId, msgId = messageId, text }, true, ct); }

    public async Task DeleteMessageAsync(string roomId, string messageId, CancellationToken ct = default)
    { using var _ = await SendAsync(HttpMethod.Post, "api/v1/chat.delete", new { roomId, msgId = messageId, asUser = true }, true, ct); }

    public async Task ReactAsync(string messageId, string emoji, bool shouldReact = true, CancellationToken ct = default)
    { using var _ = await SendAsync(HttpMethod.Post, "api/v1/chat.react", new { messageId, emoji, shouldReact }, true, ct); }

    public async Task MarkReadAsync(string roomId, CancellationToken ct = default)
    { using var _ = await SendAsync(HttpMethod.Post, "api/v1/subscriptions.read", new { rid = roomId }, true, ct); }

    public async Task<IReadOnlyList<ChatMessage>> SearchMessagesAsync(string roomId, string text, CancellationToken ct = default)
    {
        using var doc = await SendAsync(HttpMethod.Get, $"api/v1/chat.search?roomId={E(roomId)}&searchText={E(text)}&count=100", null, true, ct);
        if (!doc.RootElement.TryGetProperty("messages", out var messages)) return [];
        return messages.EnumerateArray().Select(ParseMessage).OrderBy(x => x.Timestamp).ToList();
    }

    public async Task<IReadOnlyList<UserSearchResult>> SearchUsersAsync(string term, CancellationToken ct = default)
    {
        var selector = JsonSerializer.Serialize(new { term });
        using var doc = await SendAsync(HttpMethod.Get, $"api/v1/users.autocomplete?selector={E(selector)}", null, true, ct);
        var items = doc.RootElement.TryGetProperty("items", out var i) ? i : doc.RootElement.TryGetProperty("users", out var u) ? u : default;
        if (items.ValueKind != JsonValueKind.Array) return [];
        return items.EnumerateArray().Select(x => new UserSearchResult(GetString(x, "_id"), GetString(x, "username"), GetString(x, "name"), GetString(x, "status"))).ToList();
    }

    public async Task<string> CreateDirectMessageAsync(string username, CancellationToken ct = default)
    {
        using var doc = await SendAsync(HttpMethod.Post, "api/v1/im.create", new { username }, true, ct);
        return GetString(doc.RootElement.GetProperty("room"), "rid");
    }

    public async Task<string> CreateChannelAsync(string name, bool isPrivate, CancellationToken ct = default)
    {
        var endpoint = isPrivate ? "groups.create" : "channels.create";
        using var doc = await SendAsync(HttpMethod.Post, $"api/v1/{endpoint}", new { name, members = Array.Empty<string>() }, true, ct);
        var key = isPrivate ? "group" : "channel";
        return GetString(doc.RootElement.GetProperty(key), "_id");
    }

    public async Task UploadFileAsync(string roomId, string filePath, string? description = null, CancellationToken ct = default)
    {
        EnsureAuthenticated();
        await using var stream = File.OpenRead(filePath);
        using var form = new MultipartFormDataContent();
        var file = new StreamContent(stream);
        file.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
        form.Add(file, "file", Path.GetFileName(filePath));
        if (!string.IsNullOrWhiteSpace(description)) form.Add(new StringContent(description), "description");
        using var request = NewRequest(HttpMethod.Post, $"api/v1/rooms.upload/{E(roomId)}", true);
        request.Content = form;
        using var response = await _http.SendAsync(request, ct);
        await EnsureSuccessAsync(response, ct);
    }

    private async Task<JsonDocument> SendAsync(HttpMethod method, string path, object? body, bool auth, CancellationToken ct)
    {
        using var request = NewRequest(method, path, auth);
        if (body is not null) request.Content = JsonContent.Create(body, options: _json);
        using var response = await _http.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, ct);
        return await EnsureSuccessAsync(response, ct);
    }

    private HttpRequestMessage NewRequest(HttpMethod method, string path, bool auth)
    {
        if (ServerUri is null) throw new InvalidOperationException("尚未配置服务器。");
        var request = new HttpRequestMessage(method, new Uri(ServerUri, path));
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        request.Headers.UserAgent.ParseAdd("RocketChat-CSharp/1.0");
        if (auth)
        {
            EnsureAuthenticated();
            request.Headers.Add("X-User-Id", Session!.UserId);
            request.Headers.Add("X-Auth-Token", Session.AuthToken);
        }
        return request;
    }

    private static async Task<JsonDocument> EnsureSuccessAsync(HttpResponseMessage response, CancellationToken ct)
    {
        var content = await response.Content.ReadAsStringAsync(ct);
        if (response.IsSuccessStatusCode)
            return JsonDocument.Parse(string.IsNullOrWhiteSpace(content) ? "{}" : content);
        string message = $"服务器返回 {(int)response.StatusCode} {response.ReasonPhrase}"; string? type = null;
        try { using var error = JsonDocument.Parse(content); message = GetString(error.RootElement, "message", GetString(error.RootElement, "error", message)); type = GetString(error.RootElement, "errorType"); } catch { }
        if (response.StatusCode == HttpStatusCode.Unauthorized) message = "登录失败或会话已过期，请检查账号、密码和服务器地址。";
        throw new RocketChatException(message, type);
    }

    private void ConfigureServer(string server)
    {
        if (!Uri.TryCreate(server.Trim(), UriKind.Absolute, out var uri) || (uri.Scheme != "http" && uri.Scheme != "https"))
            throw new RocketChatException("请输入有效的 http:// 或 https:// 服务器地址。");
        ServerUri = new Uri(uri.GetLeftPart(UriPartial.Authority) + uri.AbsolutePath.TrimEnd('/') + "/");
    }

    private void EnsureAuthenticated() { if (Session is null) throw new RocketChatException("请先登录。"); }
    private static string E(string value) => Uri.EscapeDataString(value);
    internal static string GetString(JsonElement e, string name, string fallback = "") => e.ValueKind == JsonValueKind.Object && e.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() ?? fallback : fallback;
    private static int GetInt(JsonElement e, string name) => e.TryGetProperty(name, out var v) && v.TryGetInt32(out var n) ? n : 0;
    private static bool GetBool(JsonElement e, string name, bool fallback = false) => e.TryGetProperty(name, out var v) && (v.ValueKind == JsonValueKind.True || v.ValueKind == JsonValueKind.False) ? v.GetBoolean() : fallback;
    private static DateTimeOffset? GetDate(JsonElement e, string name) => e.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.String && DateTimeOffset.TryParse(v.GetString(), out var d) ? d : null;
    internal static ChatMessage ParseMessage(JsonElement x)
    {
        var u = x.TryGetProperty("u", out var user) ? user : default;
        return new ChatMessage { Id = GetString(x, "_id"), RoomId = GetString(x, "rid"), UserId = GetString(u, "_id"), Username = GetString(u, "username"), DisplayName = GetString(u, "name"), Text = GetString(x, "msg"), Timestamp = GetDate(x, "ts") ?? DateTimeOffset.Now, Edited = x.TryGetProperty("editedAt", out _), IsSystem = !string.IsNullOrEmpty(GetString(x, "t")), ThreadId = GetString(x, "tmid"), ReplyCount = GetInt(x, "tcount") };
    }
    public void Dispose() => _http.Dispose();
}

File: lib/services/rocket_chat_api.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../models.dart';

class RocketChatApi {
  RocketChatApi({http.Client? client}) : _client = client ?? http.Client();
  final http.Client _client;
  Uri? serverUri;
  Session? session;

  Future<({String name, String version})> serverInfo(String server) async {
    _configureServer(server);
    final json = await _request('GET', 'api/info', authenticated: false);
    final info = json['info'] is Map ? _map(json['info']) : json;
    return (
      name: _string(info, 'name', serverUri!.host),
      version: _string(info, 'version'),
    );
  }

  Future<Session> login(String server, String username, String password) async {
    _configureServer(server);
    final json = await _request(
      'POST',
      'api/v1/login',
      body: {'user': username, 'password': password},
      authenticated: false,
    );
    final data = _map(json['data']);
    final me = _map(data['me']);
    session = Session(
      userId: _string(data, 'userId'),
      authToken: _string(data, 'authToken'),
      username: _string(me, 'username', username),
      displayName: _string(me, 'name', username),
    );
    return session!;
  }

  Future<void> logout() async {
    try {
      if (session != null) await _request('POST', 'api/v1/logout', body: {});
    } finally {
      session = null;
    }
  }

  Future<List<Room>> rooms() async {
    final json = await _request('GET', 'api/v1/subscriptions.get');
    final source = json['update'] ?? json['subscriptions'];
    if (source is! List) return [];
    final result = source
        .whereType<Map>()
        .map((raw) => _map(raw))
        .where((item) => item['open'] != false)
        .map(
          (item) => Room(
            id: _string(item, 'rid'),
            name: _string(item, 'name'),
            displayName: _string(item, 'fname', _string(item, 'name')),
            type: _string(item, 't', 'c'),
            unread: _integer(item, 'unread'),
            favorite: item['f'] == true,
            lastSeen: _date(item['ls']),
          ),
        )
        .where((room) => room.id.isNotEmpty)
        .toList();
    result.sort((a, b) {
      if (a.favorite != b.favorite) return a.favorite ? -1 : 1;
      return (b.lastSeen ?? DateTime(1970)).compareTo(
        a.lastSeen ?? DateTime(1970),
      );
    });
    return result;
  }

  Future<List<ChatMessage>> history(Room room, {int count = 100}) async {
    final endpoint = switch (room.type) {
      'd' => 'im.history',
      'p' => 'groups.history',
      _ => 'channels.history',
    };
    final json = await _request(
      'GET',
      'api/v1/$endpoint',
      query: {'roomId': room.id, 'count': count.clamp(1, 100).toString()},
    );
    return _messageList(json['messages']);
  }

  Future<ChatMessage> sendMessage(
    String roomId,
    String text, {
    String? threadId,
  }) async {
    final message = <String, dynamic>{'rid': roomId, 'msg': text};
    if (threadId != null) message['tmid'] = threadId;
    final json = await _request(
      'POST',
      'api/v1/chat.sendMessage',
      body: {'message': message},
    );
    return parseMessage(_map(json['message']));
  }

  Future<void> markRead(String roomId) async =>
      _request('POST', 'api/v1/subscriptions.read', body: {'rid': roomId});
  Future<void> updateMessage(
    String roomId,
    String messageId,
    String text,
  ) async => _request(
    'POST',
    'api/v1/chat.update',
    body: {'roomId': roomId, 'msgId': messageId, 'text': text},
  );
  Future<void> deleteMessage(String roomId, String messageId) async => _request(
    'POST',
    'api/v1/chat.delete',
    body: {'roomId': roomId, 'msgId': messageId, 'asUser': true},
  );
  Future<void> react(String messageId, {String emoji = ':+1:'}) async =>
      _request(
        'POST',
        'api/v1/chat.react',
        body: {'messageId': messageId, 'emoji': emoji, 'shouldReact': true},
      );

  Future<List<ChatMessage>> searchMessages(String roomId, String term) async {
    final json = await _request(
      'GET',
      'api/v1/chat.search',
      query: {'roomId': roomId, 'searchText': term, 'count': '100'},
    );
    return _messageList(json['messages']);
  }

  Future<List<UserResult>> searchUsers(String term) async {
    final json = await _request(
      'GET',
      'api/v1/users.autocomplete',
      query: {
        'selector': jsonEncode({'term': term}),
      },
    );
    final source = json['items'] ?? json['users'];
    if (source is! List) return [];
    return source.whereType<Map>().map((raw) {
      final item = _map(raw);
      return UserResult(
        id: _string(item, '_id'),
        username: _string(item, 'username'),
        name: _string(item, 'name'),
        status: _string(item, 'status'),
      );
    }).toList();
  }

  Future<String> createDirectMessage(String username) async {
    final json = await _request(
      'POST',
      'api/v1/im.create',
      body: {'username': username},
    );
    return _string(_map(json['room']), 'rid');
  }

  Future<String> createChannel(String name, {bool private = false}) async {
    final json = await _request(
      'POST',
      'api/v1/${private ? 'groups.create' : 'channels.create'}',
      body: {'name': name, 'members': <String>[]},
    );
    return _string(_map(json[private ? 'group' : 'channel']), '_id');
  }

  ChatMessage parseMessage(Map<String, dynamic> item) {
    final user = _map(item['u']);
    final thread = _string(item, 'tmid');
    return ChatMessage(
      id: _string(item, '_id'),
      roomId: _string(item, 'rid'),
      userId: _string(user, '_id'),
      username: _string(user, 'username'),
      displayName: _string(user, 'name'),
      text: _string(item, 'msg'),
      timestamp: _date(item['ts']) ?? DateTime.now().toUtc(),
      edited: item.containsKey('editedAt'),
      system: _string(item, 't').isNotEmpty,
      threadId: thread.isEmpty ? null : thread,
      replyCount: _integer(item, 'tcount'),
    );
  }

  List<ChatMessage> _messageList(Object? source) {
    if (source is! List) return [];
    final result = source
        .whereType<Map>()
        .map((item) => parseMessage(_map(item)))
        .toList();
    result.sort((a, b) => a.timestamp.compareTo(b.timestamp));
    return result;
  }

  Future<Map<String, dynamic>> _request(
    String method,
    String path, {
    Map<String, dynamic>? body,
    Map<String, String>? query,
    bool authenticated = true,
  }) async {
    if (serverUri == null) throw const RocketChatException('尚未配置服务器。');
    var uri = serverUri!.resolve(path);
    if (query != null) uri = uri.replace(queryParameters: query);
    final headers = <String, String>{'accept': 'application/json'};
    if (body != null) headers['content-type'] = 'application/json';
    if (authenticated) {
      if (session == null) throw const RocketChatException('请先登录。');
      headers['X-User-Id'] = session!.userId;
      headers['X-Auth-Token'] = session!.authToken;
    }
    final request = http.Request(method, uri)..headers.addAll(headers);
    if (body != null) request.body = jsonEncode(body);
    final response = await _client
        .send(request)
        .then(http.Response.fromStream)
        .timeout(const Duration(seconds: 30));
    Map<String, dynamic> json = {};
    if (response.body.isNotEmpty) {
      final decoded = jsonDecode(response.body);
      if (decoded is Map) json = _map(decoded);
    }
    if (response.statusCode >= 200 && response.statusCode < 300) return json;
    if (response.statusCode == 401) {
      throw const RocketChatException('登录失败或会话已过期，请检查账号、密码和服务器地址。');
    }
    throw RocketChatException(
      _string(
        json,
        'message',
        _string(json, 'error', '服务器返回 ${response.statusCode}。'),
      ),
      _string(json, 'errorType'),
    );
  }

  void _configureServer(String server) {
    final input = Uri.tryParse(server.trim());
    if (input == null ||
        !input.hasScheme ||
        !input.hasAuthority ||
        !{'http', 'https'}.contains(input.scheme)) {
      throw const RocketChatException('请输入有效的 http:// 或 https:// 服务器地址。');
    }
    serverUri = input.replace(
      path: '${input.path.replaceFirst(RegExp(r'/+$'), '')}/',
      query: null,
      fragment: null,
    );
  }

  void dispose() => _client.close();
  static Map<String, dynamic> _map(Object? value) =>
      value is Map ? Map<String, dynamic>.from(value) : <String, dynamic>{};
  static String _string(
    Map<String, dynamic> map,
    String key, [
    String fallback = '',
  ]) => map[key] is String ? map[key] as String : fallback;
  static int _integer(Map<String, dynamic> map, String key) =>
      map[key] is num ? (map[key] as num).toInt() : 0;
  static DateTime? _date(Object? value) =>
      value is String ? DateTime.tryParse(value) : null;
}
